File: JavaLibrary/src/test/java/com/flat/simplex/parser/logic/block/BlockIfTest.java
package com.flat.simplex.parser.logic.block;

import com.flat.simplex.parser.logic.Context;
import com.flat.simplex.parser.logic.error.Error;
import com.flat.simplex.support.TokenChain;
import org.junit.jupiter.api.Test;

import static com.flat.simplex.support.ContextSupport.assertErrors;
import static com.flat.simplex.support.TokenChain.mChain;
import static com.flat.simplex.support.TokenChain.readChain;
import static org.junit.jupiter.api.Assertions.*;

class BlockIfTest {

    @Test
    public void block() {
        TokenChain chain = readChain("if (true) hello = true;");

        Context context = new Context();
        BlockIf block = new BlockIf(context, null, chain.get(), null);
        block.read();

        assertFalse(block.isCommandBlock());
        TokenChain.assertOne("(true)", block.getTokenCondition(), "Invalid condition");
        TokenChain.assertChain("hello = true;", block.getTokenContent(), block.getTokenContentEnd(), "Invalid body");

        assertErrors(context);
    }

    @Test
    public void blockIfBlock() {
        TokenChain chain = readChain("if (true) {}");

        Context context = new Context();
        BlockIf block = new BlockIf(context, null, chain.get(), null);
        block.read();

        assertTrue(block.isCommandBlock());
        TokenChain.assertOne("(true)", block.getTokenCondition(), "Invalid condition");
        TokenChain.assertOne("{}", block.getTokenContent(), "Invalid body");

        assertErrors(context);
    }

    @Test
    public void blockIfUnexpectedToken_Fail() {
        TokenChain chain = readChain("if a (true) {}");

        Context context = new Context();
        BlockIf block = new BlockIf(context, null, chain.get(), null);
        block.read();

        assertTrue(block.isCommandBlock());
        TokenChain.assertOne("(true)", block.getTokenCondition(), "Invalid condition");
        TokenChain.assertOne("{}", block.getTokenContent(), "Invalid body");

        assertErrors(context, Error.unexpectedToken);
    }

    @Test
    public void blockIfUnexpectedEndOfTokens_Fail() {
        TokenChain chain = readChain("if (true)");

        Context context = new Context();
        BlockIf block = new BlockIf(context, null, chain.get(), null);
        block.read();

        assertFalse(block.isCommandBlock());
        TokenChain.assertOne("(true)", block.getTokenCondition(), "Invalid condition");

        assertErrors(context, Error.unexpectedEndOfTokens);
    }

    @Test
    public void blockIfMissingCondition_Fail() {
        TokenChain chain = readChain("if () hello = true;");

        Context context = new Context();
        BlockIf block = new BlockIf(context, null, chain.get(), null);
        block.read();

        assertFalse(block.isCommandBlock());
        TokenChain.assertOne("()", block.getTokenCondition(), "Invalid condition");
        TokenChain.assertChain("hello = true;", block.getTokenContent(), block.getTokenContentEnd(), "Invalid body");

        assertErrors(context, Error.ifConditionExpected);
    }

    @Test
    public void blockIfMissingCloser_Fail() {
        TokenChain chain = readChain("if (true) {");

        Context context = new Context();
        BlockIf block = new BlockIf(context, null, chain.get(), null);
        block.read();

        assertTrue(block.isCommandBlock());
        TokenChain.assertOne("(true)", block.getTokenCondition(), "Invalid condition");
        TokenChain.assertOne("{", block.getTokenContent(), "Invalid body");

        assertErrors(context, Error.missingCloser);
    }
}

File: SimplexApp/simplex/lang/parser/logic/block/BlockIf.test.js
const TokenChain = require("simplex/lang/support/TokenChain.js");
const Context = require("simplex/lang/parser/logic/Context.js");
const Error = require("simplex/lang/parser/logic/error/Error.js");
const BlockIf = require("simplex/lang/parser/logic/block/BlockIf.js")
const ContextSupport = require("simplex/lang/support/ContextSupport.js");
const AssertError = require("simplex/lang/support/AssertError.js");

const readChain = TokenChain.readChain;
const assertErrors = ContextSupport.assertErrors;
const assertTrue = AssertError.assertTrue;
const assertFalse = AssertError.assertFalse;

test('block', () => {
    let chain = readChain('if (true) hello = true;');
    let context = new Context(chain.get());
    let block = new BlockIf(context, chain.get(), null);
    block.read();
    assertFalse(block.isCommandBlock());
    TokenChain.assertOne('(true)', block.getTokenCondition(), 'Invalid condition');
    TokenChain.assertChain('hello = true;', block.getTokenContent(), block.getTokenContentEnd(), 'Invalid body');
    assertErrors(context);
});
test('blockIfBlock', () => {
    let chain = readChain('if (true) {}');
    let context = new Context(chain.get());
    let block = new BlockIf(context, chain.get(), null);
    block.read();
    assertTrue(block.isCommandBlock());
    TokenChain.assertOne('(true)', block.getTokenCondition(), 'Invalid condition');
    TokenChain.assertOne('{}', block.getTokenContent(), 'Invalid body');
    assertErrors(context);
});
test('blockIfUnexpectedToken_Fail', () => {
    let chain = readChain('if a (true) {}');
    let context = new Context(chain.get());
    let block = new BlockIf(context, chain.get(), null);
    block.read();
    assertTrue(block.isCommandBlock());
    TokenChain.assertOne('(true)', block.getTokenCondition(), 'Invalid condition');
    TokenChain.assertOne('{}', block.getTokenContent(), 'Invalid body');
    assertErrors(context, Error.unexpectedToken);
});
test('blockIfUnexpectedEndOfTokens_Fail', () => {
    let chain = readChain('if (true)');
    let context = new Context(chain.get());
    let block = new BlockIf(context, chain.get(), null);
    block.read();
    assertFalse(block.isCommandBlock());
    TokenChain.assertOne('(true)', block.getTokenCondition(), 'Invalid condition');
    assertErrors(context, Error.unexpectedEndOfTokens);
});
test('blockIfMissingCondition_Fail', () => {
    let chain = readChain('if () hello = true;');
    let context = new Context(chain.get());
    let block = new BlockIf(context, chain.get(), null);
    block.read();
    assertFalse(block.isCommandBlock());
    TokenChain.assertOne('()', block.getTokenCondition(), 'Invalid condition');
    TokenChain.assertChain('hello = true;', block.getTokenContent(), block.getTokenContentEnd(), 'Invalid body');
    assertErrors(context, Error.ifConditionExpected);
});
test('blockIfMissingCloser_Fail', () => {
    let chain = readChain('if (true) {');
    let context = new Context(chain.get());
    let block = new BlockIf(context, chain.get(), null);
    block.read();
    assertTrue(block.isCommandBlock());
    TokenChain.assertOne('(true)', block.getTokenCondition(), 'Invalid condition');
    TokenChain.assertOne('{', block.getTokenContent(), 'Invalid body');
    assertErrors(context, Error.missingCloser);
});